---
title: "Consensus frequency matrix compared with quasibam"
author: "Stef Piatek"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  html_document:
    keep_md: TRUE
params:
  date_prefix: "170908"
---
Script to compare the frequency matrix from the FASTA pileup to quasibam from the same consensus

Example usage from command line:

    Rscript -e "rmarkdown::render('scripts/frequency-matrix_comparison.Rmd', params = list(pipeline = 'vicuna_bwa', date_prefix = '170908') 'html_document', '../reports/170908_frequency-matrix_quasibam.html')"
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Biostrings)  # bioconductor package
library(readr)
library(here)
library(glue)
library(dplyr)
library(tidyr)
library(stringr)
library(ggplot2)
library(broom)
library(purrr)
library(forcats)


# set up ggplot theme and palettes 
theme_custom <- function (base_size = 12, base_family = "") {
  theme_bw(base_size = base_size, base_family = base_family) %+replace%
    theme(
      axis.line.x = element_line(color = "black"),
      axis.line.y = element_line(color = "black"),
      panel.border = element_blank(),
      legend.key = element_rect(fill = NA, colour = NA),
      panel.grid.minor.y = element_line(colour = "white"),
      panel.grid.major = element_line(colour = "white"),
      axis.title.x = element_text(vjust = -1.5),
      plot.margin = unit(c(0.5, 0, 0.5, 0), "cm")# top right bottom left
    )
}
theme_set(theme_custom(base_size=12))

palette_5 <- c("#FFA347","#142952", "#B80000","#97F297", "#000000")

# set up boolean to trigger vphaser chunk if files exist
vphaser_samples <- list.files(here("data", "vphaser")) %>%
  str_extract(glue("({params$date_prefix}_[0-9]+)")) %>%
  na.omit() %>%
  unique() 

do_vphaser_comparison <- if_else(length(vphaser_samples) >= 1,
                                 TRUE,
                                 FALSE)

```

## Comparison of consensus frequency matrix to quasibam

*When the majority of FASTAs have a gap (- or N), the base has been removed from the consensus sequence*

- Consensus frequency matrix made from input FASTAs for simulating quasispecies reads
- FASTQs have been generated from these FASTAs, mapped back to the consensus FASTA sequence and then quasi_bam run
- This document compares the consensus quasibam output to the pipeline.

    - The NS3, NS5A and NS5B regions have been extracted from the files, discarding       
      everything else
    
```{r load_data, warning=FALSE, echo=FALSE, message=FALSE}
# Load helper scripts and hcv reference information
source(here("scripts", "helper_functions.R"))

## Load up samples and filter them to roi ----
# get list of samples
date_prefix <- params$date_prefix
print(glue("Analysing data from {date_prefix}"))

samples <- list.files(here("data")) %>%
  str_extract(glue("({date_prefix}_[0-9]+)")) %>%
  unique() %>% 
  na.omit()


# set up initial dataframe 
consensus_fm <- read_delim(file = here("data", glue("{samples[1]}_quasi_frequency_matrix.txt")),
                           delim = "\t") %>%
  mutate(sample_name = samples[1]) %>%
  filter_to_roi()

consensus_qb <- read_delim(file = here("data", glue("{samples[1]}_quasi_sorted.txt")),
                           delim = "\t") %>%
  select(-X22) %>%  # extra column added in import
  mutate(sample_name = samples[1])%>%
  filter_to_roi()

# add all other sets to the dataframe
load_data <- function(sample){
  consensus_fm <<- read_delim(file = here("data", 
                                         glue("{sample}_quasi_frequency_matrix.txt")),
                             delim = "\t") %>%
    mutate(sample_name = sample) %>%
    filter_to_roi() %>%
    bind_rows(consensus_fm)
  
  consensus_qb <<- read_delim(file = here("data", glue("{sample}_quasi_sorted.txt")),
                           delim = "\t") %>%
    select(-X22) %>%
    mutate(sample_name = sample)%>%
    filter_to_roi() %>%
    bind_rows(consensus_qb)  
}

walk(.x=samples[2:length(samples)], .f=load_data)

```

## Discordant basecalls


```{r discordant_basecalls, echo=FALSE}

### Compare size of reference between
ref_length_consensus <- consensus_fm %>%
  group_by(sample_name, region_name) %>%
  summarise(ref_length = max(Pos))

consensus_qb %>%
  group_by(sample_name, region_name) %>%
  summarise(ref_length = max(Pos)) %>%
  full_join(ref_length_consensus, by = c("sample_name", "region_name"), suffix = c("_pipeline", "_consensus")) %>%
  filter(ref_length_consensus != ref_length_pipeline)

### Compare data where basecalls are discordant ----
discordant_basecalls <- consensus_qb %>%
  group_by(sample_name, region_name) %>%
  filter(RefN != Cons) %>%
  filter(!(Cons == "R" & RefN %in% c("A", "G"))) %>%
  filter(!(Cons == "Y" & RefN %in% c("C", "T"))) %>%
  filter(!(Cons == "S" & RefN %in% c("G", "C"))) %>%
  filter(!(Cons == "W" & RefN %in% c("A", "T"))) %>%
  filter(!(Cons == "K" & RefN %in% c("G", "T"))) %>%
  filter(!(Cons == "M" & RefN %in% c("A", "C")))

discordant_basecalls %>%
  select(Pos:Gap, Cons, sample_name, region_name) %>%
  filter(Depth != 0) %>%
  arrange(Pos)
  
```

## Basecall frequency differences

### Parse data for plotting


```{r base_frequency_data}
tmp_consensus_diff <- consensus_fm %>%
  full_join(consensus_qb, by=c("Pos", "sample_name", "region_name", "RefN"), 
            suffix=c("_consensus", "_quasibam")) %>%
  # Keep only quasibam depth as depth
  mutate(depth = Depth_quasibam) %>%
  select(- starts_with("Depth_")) %>%
  # tidy data to have base, sample and value columns
  gather(key = base_source, value = value, 
         starts_with("A_"), starts_with("C_"), starts_with("T_"), starts_with("G_"), 
         starts_with("Gap_")  ) %>%
  separate(col = base_source, into = c("base", "source"), sep = "_" ) %>%
  # get difference in percent
  group_by(Pos, base, sample_name, region_name) %>%
  arrange(source) %>%
  mutate(pc_diff = max(value) - min(value)) %>%
  # plot using log, so no negative. use factor to determine which value is greater
  mutate(consensus_greater = if_else(value > lead(value),
                                     "consensus value greater",
                                     "consensus value smaller",
                                     missing="quasibam_sample")) %>%
  mutate(consensus_greater = parse_factor(consensus_greater, 
                                          levels = c("consensus value greater",
                                                     "consensus value smaller",
                                                     "quasibam_sample"))) %>%
  ungroup() %>%
  mutate(sample_name = if_else(str_detect(sample_name, "([0-9]{6}_[0-9]{1}$)"),
                               str_replace(sample_name, "_", "_0"),
                               sample_name)) 

# reshape quality data into long format and merge back into consensus_diff
quality_values <- tmp_consensus_diff %>%
  select(Pos, starts_with("q"), sample_name, region_name) %>%
  gather(key = base, value = quality_value, starts_with("q")) %>%
  mutate(base = str_replace(base, "q", "")) %>%
  group_by(Pos, base, sample_name, region_name) %>%
  dplyr::slice(1) %>%
  ungroup()

consensus_diff <- tmp_consensus_diff %>%
  select(-starts_with("q")) %>%
  full_join(quality_values, by=c("Pos", "base", "sample_name", "region_name")) %>%
  mutate(base = parse_factor(base, levels=c("A", "C", "G", "T", "Gap"))) %>%
  # for non-called bases, change base quality to NA
  mutate(quality_value = replace(quality_value,
                                 quality_value == 0 & value == 0,
                                 NA))
# remove temporary data
rm("quality_values")
rm("tmp_consensus_diff")

```

### Plotting base frequency differences

- Histrograms have a 0.1% added to frequency difference (to plot 0 on log scale), dotplots have 1% threshold. 
- Largest base frequency differences in NS5A, otherwise above ~ 4% is quite rare

    - Most differences above 1% are at the 5' end of the region
    - Most differences above 1% have depth < 1,300: difficult to map?

- Generally, increased base frequency in consensus matrix compared to quasibam
- Sample 13 and 18 have particularily large differences in base frequency


```{r explore_data_plotting}

## Base coverage above 100 ----
consensus_qb %>%
  group_by(Pos, region_name) %>%
  filter(Depth > 100) %>%
  summarise(percent_covered = 100 * n() / length(samples)) %>%
  arrange(region_name) %>%
  ggplot() +
  geom_point(aes(x = Pos, y = percent_covered, colour = '>= 100')) + 
  scale_x_continuous(limits=c(1, max(consensus_fm$Pos))) +
  scale_colour_manual(values=palette_5) +
  facet_wrap(~region_name, scales = "free_x") + 
  labs(x = "Nucleotide position",
       y = "Percent of samples",
       title = "Percent of samples with depth above 100",
       colour = "Depth")

# double check no missed rows
consensus_qb %>%
  group_by(Pos, region_name) %>%
  summarise(percent_covered = 100 * n() / length(samples)) %>%
  group_by(region_name) %>%
  arrange(region_name) %>%
  filter(lag(Pos) != Pos - 1)

## Distribution of frequency differences per base ----
consensus_diff %>%
  filter(consensus_greater != "quasibam_sample") %>%
  filter(pc_diff >= 0.1) %>%  ggplot(aes(x=pc_diff, fill=base)) +
  geom_histogram(bins=40) +
  scale_x_log10() + 
  scale_y_log10() + 
  facet_grid(consensus_greater ~ region_name) + 
  scale_fill_manual(values=palette_5) +
  labs(x = "Frequency difference",
       y = "Count",
       title = "Differences compared to consensus")

## Distribution of frequency differences per sample ----
consensus_diff %>%
  filter(consensus_greater != "quasibam_sample") %>%
  filter(pc_diff >= 0.1) %>%
  ggplot(aes(x=pc_diff, fill=region_name)) +
  geom_histogram(bins=40) +
  scale_x_log10() + 
  scale_y_log10() + 
  scale_fill_manual(values=palette_5) +
  facet_wrap(~ sample_name) + 
  labs(x = "Frequency difference",
       y = "Count",
       fill = "Region name",
       title = "Differences compared to consensus per sample")

## Per position frequency differences ----
consensus_diff %>%
  filter(consensus_greater != "quasibam_sample") %>%
  filter(pc_diff >= 1) %>%
  ggplot(aes(x=Pos, fill=region_name)) +
  geom_histogram(bins=40) +
  scale_fill_manual(values=palette_5) +
  facet_wrap(~ sample_name, scales = "free_y") + 
  labs(x = "Nucleotide Position",
       y = "Count",
       fill = "Region name",
       title = "Number of differences >1% by position") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))

## proportion of samples with base difference at position ----
consensus_diff %>%
  filter(consensus_greater != "quasibam_sample") %>%
  mutate(greater_1_pc = if_else(pc_diff >= 1, 1, 0)) %>% 
  group_by(Pos, base, source, region_name) %>%
  summarise(gt_1_pc = sum(100 * greater_1_pc) / n(),
            max_pc_diff = max(pc_diff)) %>%
  filter(max_pc_diff >= 1) %>%
ggplot(aes(x = max_pc_diff, y = gt_1_pc, colour = base, alpha = 0.2)) +
  geom_jitter() +
  scale_x_log10() + 
  facet_grid(region_name ~ base) + 
  scale_colour_manual(values=palette_5) +
  guides(alpha=FALSE) + 
  labs(x = "Maximum difference in base frequency",
       y = "Percent of samples",
       title = "Percent of samples with >1% difference in base frequency")

## Distribution of differences greater than 1% by reference position ----
consensus_diff %>%
  # remove samples with frequencies within 1% of eachother 
  filter(pc_diff >= 1) %>%
  filter(consensus_greater != "quasibam_sample") %>%
  ggplot() +
  geom_histogram(bins=40, aes(x=Pos, fill=base)) +
  facet_grid(consensus_greater ~ region_name, scales = "free_x") + 
  scale_fill_manual(values=palette_5) +
  labs(x = "Nucleotide position",
       y = "Count",
       title = "Differences by position, over 1%") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))

## Read depth of reads with frequency differences greater than 1% ----
consensus_diff %>%
  filter(consensus_greater != "quasibam_sample") %>%
  filter(pc_diff >= 1) %>%
  ggplot() +
  geom_histogram(bins = 40, aes(x=depth, fill=base)) +
  facet_grid(consensus_greater ~ region_name) +
  scale_fill_manual(values=palette_5) +
  labs(x = "Read Depth",
       y = "Count",
       title = "Read depth of frequency differences >= 1%") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))

## Per sample distribution by nucleotide position
consensus_diff %>%
  mutate(greater_1_pc = if_else(pc_diff >= 1, 1, 0)) %>% 
  filter(quality_value > 1) %>%
  filter(source == "consensus") %>%
  group_by(Pos, region_name) %>%
  mutate(gt_1_pc = sum(100 * greater_1_pc) / n()) %>%
  ungroup() %>%
  group_by(Pos, region_name) %>%
  filter(pc_diff >= 1) %>%
  filter(pc_diff == max(pc_diff)) %>%  # Keep maximum diff per position, not per base & position
  ggplot() +
  geom_point(stat = "identity", aes(x = Pos, y = pc_diff, alpha = 0.2, 
                                    colour = gt_1_pc)) +
  scale_y_log10() +
  facet_wrap( ~ sample_name, scales = "free_x") + 
  scale_color_continuous(low = palette_5[1] , high = palette_5[5]) +
  guides(alpha = FALSE) +
  labs(x = "Nucleotide Position",
       y = "Maximum percent difference per base",
       colour = "Percent bases\nabove 1%") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))

```

## Number of positions comparison

- for 171009, samples 10, 11 and 12 have been filtered out

```{r filtered_comparison, echo=FALSE, results='asis'}


total_bases <- consensus_qb %>%
  group_by(region_name, sample_name) %>%
  summarise(max_length = max(Pos) * 4)  # as 4 bases

total_bases_filtered <- consensus_qb %>%
  filter(!sample_name %in% c('171009_10', '171009_11', '171009_12')) %>%
  group_by(region_name, sample_name) %>%
  summarise(max_length = max(Pos) * 4)  # as 4 bases

walk(.x = c(1, 5, 10),
     .f = analyse_threshold,
     source_name = "consensus")

pass_filter %>%
  filter(!sample_name %in% c('171009_10', '171009_11', '171009_12')) %>%
  group_by(Pos, region_name) %>%
  filter(pc_diff == max(pc_diff)) %>%
  select(Pos:region_name, Cons, depth:pc_diff, -source, base) %>%
  ungroup() %>%
  arrange(Pos, sample_name, region_name) %>%
  knitr::kable(caption = "All bases with greater than 10% difference")

```

## Vphaser comparison

- If vphaser analysis has been carried out, this chunk will run
- Vphaser failed on some samples so only passing samples are compared
- Vphaser assumes consensus sequence used is perfect and calculates the percent from 100% of the reference base and 0 % 
  for non-reference bases. 

```{r vphaser, warning=FALSE, message=FALSE, echo=FALSE, results='asis', eval=do_vphaser_comparison}

## Functions for loading vphaser data ----

filter_by_position <- function(start, end, region, table_to_filter){
  # filter table to given start and end position for ROI mapping
   table_to_filter %>%
    mutate(Pos = as.numeric(Pos)) %>%
    filter(Pos >= start & Pos <= end) %>%
    mutate(region_name = region) %>%
    mutate(region_position = Pos + 1 - start) %>%
    return()
}

load_vphaser <- function(vphaser_sample){
  # loads individual vphaser sample and selects only roi positions
  consensus_fasta <- read_delim(file = here("data", 
                                            glue("{vphaser_sample}_quasi_consensus.fas")), 
                        delim = "\t", col_names = "consensus") %>%
    filter(!startsWith(consensus, ">")) 
  consensus_seq <- str_c(consensus_fasta$consensus, collapse="")

  roi_alignment <- pairwiseAlignment(pattern = roi$sequence, 
                                     subject = consensus_seq,
                                     type = "global-local")
  # tibble to return
   sample_vphaser <- read_delim(file = here("data", "vphaser", vphaser_sample,
                                           "consensus.fdr.var.txt"),
                                delim = "\t") %>%
     mutate(sample_name = vphaser_sample) %>%
     rename(Pos = `# Ref_Pos`,
            RefN = Cons) %>%
     filter(!str_detect(Pos, '#.*')) %>%
     pmap_df(.l = list(start = start(subject(roi_alignment)),
                       end = end(subject(roi_alignment)),
                       region = roi$region_name),
             table_to_filter = .,
             .f = filter_by_position) %>%
     mutate(sample_name = if_else(str_detect(sample_name, "([0-9]{6}_[0-9]{1}$)"),
                               str_replace(sample_name, "_", "_0"),
                               sample_name))
     
     return(sample_vphaser)
}

load_all_vphaser <- function(vphaser_sample){
  # loads and merges all vphaser sampels
  vphaser <<- load_vphaser(vphaser_sample) %>%
    bind_rows(vphaser)
}

## Load samples that passed vphaser ----
vphaser <- load_vphaser(vphaser_samples[1])

walk(.x=vphaser_samples[2:length(vphaser_samples)], 
     .f=load_all_vphaser)

## compare vphaser and quasibam ----
qb_vphaser <- vphaser %>%
  inner_join(consensus_diff, 
             by = c("region_position" = "Pos", "Var" = "base", 
                    "region_name", "sample_name", "RefN"),
             suffix = c("_vphaser", "_quasibam")) %>%
  # calculate actual value difference from frequency matrix values
  filter(source == "consensus") %>%
  mutate(fm_vphaser_pc_diff = abs(Var_perc - value)) %>%
  # now look at the relative difference from quasibam and vphaser
  mutate(qb_vphaser_pc_diff = pc_diff - fm_vphaser_pc_diff) %>%
  mutate(greater_pc_diff = if_else(fm_vphaser_pc_diff > pc_diff,
                                   "Vphaser greater",
                                   "Quasibam greater"))

ggplot(qb_vphaser, aes(x = abs(qb_vphaser_pc_diff), fill = greater_pc_diff)) +
  geom_histogram(position="identity", bins = 40, alpha=0.6) +
  scale_x_log10(
    labels = scales::trans_format("log10", scales::math_format(10^.x))
  ) +
  scale_fill_manual(values = palette_5) +
  annotation_logticks(sides = "b") +
  labs(x = "Percent difference by base",
       y = "Count",
       fill = "Greater difference\nto true value",
       title = "Absolute difference between quasibam and vphaser") 

## Number of positions different using thresholds----
vphaser_position_threshold <- function(diff_threshold, pval_threshold){
  output <- vphaser %>%
    group_by(Pos) %>%
    filter(Var_perc == max(Var_perc)) %>%
    filter(Var_perc > diff_threshold) %>%
    filter(Strd_bias_pval < pval_threshold) %>%
    group_by(region_name) %>%
    summarise(`Total number` = n())
  
  if(length(output$region_name > 0)){
    output %>%
      knitr::kable(caption = glue(
        "Vphaser differences > {diff_threshold} % with P < {pval_threshold}")
      ) %>%
    print()
  }
}


walk(.x = c(1, 5, 10),
     .f = vphaser_position_threshold,
     pval_threshold = 1)

walk(.x = c(1, 5, 10),
     .f = vphaser_position_threshold,
     pval_threshold = 0.05)  
```

### Positions without variation across all samples

- Used minimum value of 1% to count as true difference
- Smoothed line is shown with 95% confidence interals in grey
- Behind this the data range is shown in colour

```{r no_variation, echo=FALSE, eval=do_vphaser_comparison}


max_consensus_diff <- consensus_diff %>%
  group_by(sample_name, Pos, region_name) %>%
  filter(source == 'consensus') %>%
  filter(pc_diff == max(pc_diff)) %>%
  # because bases with no difference will be maximum
  # get first in group to only have one base
  dplyr::slice(1) %>%
  mutate(quasibam_no_difference = if_else(pc_diff <= 1,
                                          1,
                                          0)) %>%
  group_by(region_name, Pos) %>%
  mutate(percent_quasibam_absent = 100 * sum(quasibam_no_difference) / n()) %>%
  select(-quasibam_no_difference)

full_qb_vphaser <- vphaser %>%
  filter(Var_perc >= 1) %>%
  full_join(max_consensus_diff, 
             by = c("region_position" = "Pos",  
                    "region_name", "sample_name"),
             suffix = c("_vphaser", "_quasibam")) %>%
  filter(is.na(Strd_bias_pval)) %>%
  filter(sample_name %in% unique(vphaser$sample_name)) %>%
  select(sample_name:percent_quasibam_absent) %>%
  group_by(region_position, region_name) %>%
  mutate(percent_vphaser_absent = 100 * n() / length(vphaser_samples)) %>%
  gather(key = "variation_caller", value = "percent_absent", 
         ends_with("absent"), factor_key = TRUE) %>%
  mutate(variation_caller = fct_recode(variation_caller, 
                                       Quasibam = "percent_quasibam_absent",
                                       Vphaser = "percent_vphaser_absent"))
  
ggplot(full_qb_vphaser, aes(x = region_position, y = percent_absent, 
                            colour = variation_caller)) +  
  geom_line(alpha = 0.2) +
  geom_smooth() + 
  facet_wrap(~ region_name, scales = "free_x") +
  labs(x = "Position (nt)",
       y = "Percent of samples with < 1% variation",
       colour = "Variation caller",
       title = "Positions with less than 1% variation") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))
```

