---
title: "Quasibam comparison"
author: "Stef Piatek"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Biostrings)  # bioconductor package
library(readr)
library(here)
library(glue)
library(dplyr)
library(tidyr)
library(stringr)
library(ggplot2)
library(broom)

# set up ggplot theme and palettes 
theme_custom <- function (base_size = 12, base_family = "") {
  theme_bw(base_size = base_size, base_family = base_family) %+replace%
    theme(
      axis.line.x = element_line(color = "black"),
      axis.line.y = element_line(color = "black"),
      panel.border = element_blank(),
      legend.key = element_rect(fill = NA, colour = NA),
      panel.grid.minor.y = element_line(colour="white"),
      panel.grid.major = element_line(colour = "white"),
      axis.title.x=element_text(vjust=-1.5),
      plot.margin= unit(c(0.5, 0, 0.5, 0), "cm")# top right bottom left
    )
}
theme_set(theme_custom(base_size=12))

palette_5 <- c("#FFA347","#142952", "#B80000","#97F297", "#000000")


```

## Comparison of consensus frequency matrix to quasibam

*When the majority of FASTAs have a gap (- or N), the base has been removed from the consensus sequence*

- Consensus frequency matrix made from input FASTAs for simulating quasispecies reads
- FASTQs have been generated from these FASTAs, mapped back to the consensus FASTA sequence and then quasi_bam run
- This document compares the consensus quasibam output to the pipeline.

    - The NS3, NS5A and NS5B regions have been extracted from the files, discarding       everything else
    
```{r load_data, warning=FALSE, echo=FALSE, message=FALSE}
# Load helper scripts and hcv reference information
source(here("scripts", "helper_functions.R"))


## Load up samples and filter them to roi ----
# get list of samples
samples <- list.files(here("data")) %>%
  str_extract("([0-9]{6}_[0-9]+)") %>%
  unique() %>% 
  na.omit()


# set up initial dataframe 
consensus_fm <- read_delim(file = here("data", glue("{samples[1]}_quasi_frequency_matrix.txt")),
                           delim = "\t") %>%
  mutate(sample_name = samples[1]) %>%
  filter_to_roi()



consensus_qb <- read_delim(file = here("data", glue("{samples[1]}_quasi_sorted.txt")),
                           delim = "\t") %>%
  select(-X22) %>%  # extra column added in import
  mutate(sample_name = samples[1])%>%
  filter_to_roi()


# add all other sets to the dataframe
for(item in 2:length(samples)){
  sample <- samples[item]
  consensus_fm <- read_delim(file = here("data", glue("{sample}_quasi_frequency_matrix.txt")),
                           delim = "\t") %>%
    mutate(sample_name = sample) %>%
    filter_to_roi() %>%
    bind_rows(consensus_fm)
  
  consensus_qb <- read_delim(file = here("data", glue("{sample}_quasi_sorted.txt")),
                           delim = "\t") %>%
    select(-X22) %>%
    mutate(sample_name = sample)%>%
    filter_to_roi() %>%
    bind_rows(consensus_qb)  
}

```

## Discordant basecalls

- Only discordant basecalls is where multiple bases could be called, or there is a gap, likely due to low depth

```{r discordant_basecalls, echo=FALSE}

### Compare size of reference between
ref_length_consensus <- consensus_fm %>%
  group_by(sample_name, region_name) %>%
  summarise(ref_length = max(Pos))

consensus_qb %>%
  group_by(sample_name, region_name) %>%
  summarise(ref_length = max(Pos)) %>%
  full_join(ref_length_consensus, by = c("sample_name", "region_name"), suffix = c("_pipeline", "_consensus")) %>%
  filter(ref_length_consensus != ref_length_pipeline)


### Compare data where basecalls are discordant ----

discordant_basecalls <- consensus_qb %>%
  group_by(sample_name, region_name) %>%
  filter(RefN != Cons) %>%
  filter(!(Cons == "R" & RefN %in% c("A", "G"))) %>%
  filter(!(Cons == "Y" & RefN %in% c("C", "T"))) %>%
  filter(!(Cons == "S" & RefN %in% c("G", "C"))) %>%
  filter(!(Cons == "W" & RefN %in% c("A", "T"))) %>%
  filter(!(Cons == "K" & RefN %in% c("G", "T"))) %>%
  filter(!(Cons == "M" & RefN %in% c("A", "C")))
  

discordant_basecalls %>%
  select(Pos:Gap, Cons, sample_name, region_name) %>%
  filter(Depth != 0) %>%
  arrange(Pos)
  
```

## Basecall frequency differences

### Parse data for plotting


```{r base_frequency_data}
tmp_consensus_diff <- consensus_fm %>%
  full_join(consensus_qb, by=c("Pos", "sample_name", "region_name", "RefN"), 
            suffix=c("_consensus", "_quasibam")) %>%
  # keep only rows with 200 read depth -- is this reasonable?
  filter(Depth_quasibam >= 200) %>%
  # Keep only quasibam depth as depth
  mutate(depth = Depth_quasibam) %>%
  select(- starts_with("Depth_")) %>%
  # tidy data to have base, sample and value columns
  gather(key = base_source, value = value, 
         starts_with("A_"), starts_with("C_"), starts_with("T_"), starts_with("G_"), 
         starts_with("Gap_")  ) %>%
  separate(col = base_source, into = c("base", "source"), sep = "_" ) %>%
  # get difference in percent
  group_by(Pos, base, sample_name, region_name) %>%
  arrange(source) %>%
  mutate(pc_diff = max(value) - min(value)) %>%
  # plot using log, so no negative. use factor to determine which value is greater
  mutate(consensus_greater = if_else(value > lead(value),
                                     "consensus value greater",
                                     "consensus value smaller",
                                     missing="quasibam_sample")) %>%
  mutate(consensus_greater = parse_factor(consensus_greater, 
                                          levels = c("consensus value greater",
                                                     "consensus value smaller",
                                                     "quasibam_sample"))) %>%
  ungroup() %>%
  mutate(sample_name = if_else(str_detect(sample_name, "([0-9]{6}_[0-9]{1}$)"),
                               str_replace(sample_name, "_", "_0"),
                               sample_name)) 



# reshape quality data into long format and merge back into consensus_diff
quality_values <- tmp_consensus_diff %>%
  select(Pos, starts_with("q"), sample_name, region_name) %>%
  gather(key = base, value = quality_value, starts_with("q")) %>%
  mutate(base = str_replace(base, "q", "")) %>%
  group_by(Pos, base, sample_name, region_name) %>%
  dplyr::slice(1) %>%
  ungroup()

consensus_diff <- tmp_consensus_diff %>%
  select(-starts_with("q")) %>%
  full_join(quality_values, by=c("Pos", "base", "sample_name", "region_name")) %>%
  mutate(base = parse_factor(base, levels=c("A", "C", "G", "T", "Gap"))) %>%
  # for non-called bases, change base quality to NA
  mutate(quality_value = replace(quality_value,
                                 quality_value == 0 & value == 0,
                                 NA))
# remove temporary data
rm("quality_values")
rm("tmp_consensus_diff")

```

### Plotting base frequency differences

- Histrograms have a 0.1% added to frequency difference (to plot 0 on log scale), dotplots have 1% threshold. 
- Largest base frequency differences in NS5A, otherwise above ~ 4% is quite rare

    - Most differences above 1% are at the 5' end of the region
    - Most differences above 1% have depth < 1,300: difficult to map?

- Generally, increased base frequency in consensus matrix compared to quasibam
- Sample 13 and 18 have particularily large differences in base frequency


```{r explore_data_plotting}

## Bases with coverage of 200 ----
consensus_qb %>%
  group_by(Pos, region_name) %>%
  summarise(percent_covered = 100 * n() / length(samples)) %>%
  arrange(region_name) %>%
  ggplot() +
  geom_point(aes(x = Pos, y = percent_covered, colour = '>= 200')) + 
  scale_x_continuous(limits=c(1, max(consensus_fm$Pos))) +
  scale_colour_manual(values=palette_5) +
  facet_wrap(~region_name) + 
  labs(x = "Nucleotide position",
       y = "Percent of samples",
       title = "Percent of samples with depth above 200",
       colour = "Depth")

# double check no missed rows
consensus_qb %>%
  group_by(Pos, region_name) %>%
  summarise(percent_covered = 100 * n() / length(samples)) %>%
  group_by(region_name) %>%
  arrange(region_name) %>%
  filter(lag(Pos) != Pos - 1)
# coverage looks good

## Distribution of frequency differences per base ----
consensus_diff %>%
  filter(consensus_greater != "quasibam_sample") %>%
  mutate(pc_diff = pc_diff + 0.1) %>%
  ggplot(aes(x=pc_diff, fill=base)) +
  geom_histogram(bins=40) +
  scale_x_log10() + 
  scale_y_log10() + 
  facet_grid(consensus_greater ~ region_name) + 
  scale_fill_manual(values=palette_5) +
  labs(x = "Frequency difference (with 0.1% added)",
       y = "Count",
       title = "Differences compared to consensus")
# NS5A har most errors, with lower base frequency in matrix compared to quasibam
# Generally increased base frequency in matrix compared to quasibam

## Distribution of frequency differences per sample ----
consensus_diff %>%
  filter(consensus_greater != "quasibam_sample") %>%
  mutate(pc_diff = pc_diff + 0.1) %>%
  ggplot(aes(x=pc_diff, fill=sample_name)) +
  geom_histogram(bins=40) +
  scale_x_log10() + 
  scale_y_log10() + 
  facet_wrap(~ sample_name) + 
  labs(x = "Frequency difference (with 0.1% added)",
       y = "Count",
       title = "Differences compared to consensus")
# sample 13 and 18 are partcularily bad for large differences

## proportion of samples with base difference at position ----
consensus_diff %>%
  filter(consensus_greater != "quasibam_sample") %>%
  mutate(greater_1_pc = if_else(pc_diff >= 1, 1, 0)) %>% 
  group_by(Pos, base, source, region_name) %>%
  summarise(gt_1_pc = sum(100 * greater_1_pc) / n(),
            max_pc_diff = max(pc_diff)) %>%
  filter(max_pc_diff >= 1) %>%
ggplot(aes(x = max_pc_diff, y = gt_1_pc, colour = base, alpha = 0.2)) +
  geom_jitter() +
  scale_x_log10() + 
  facet_grid(region_name ~ base) + 
  scale_colour_manual(values=palette_5) +
  guides(alpha=FALSE) + 
  labs(x = "Maximum difference in base frequency",
       y = "Percent of samples",
       title = "Percent of samples with >1% difference in base frequency")
# NS5A has most common large differences

## Distribution of differences greater than 1% by reference position ----
consensus_diff %>%
  # remove samples with frequencies within 1% of eachother 
  filter(pc_diff >= 1) %>%
  filter(consensus_greater != "quasibam_sample") %>%
  ggplot() +
  geom_histogram(bins=40, aes(x=Pos, fill=base)) +
  facet_grid(consensus_greater ~ region_name) + 
  scale_fill_manual(values=palette_5) +
  labs(x = "Nucleotide position",
       y = "Count",
       title = "Differences by position, over 1%") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))
# 5' end of NS5A has a large number of frequency differences, this may be inflated due to each base being different in one position

## Read depth of reads with frequency differences greater than 1%
consensus_diff %>%
  filter(consensus_greater != "quasibam_sample") %>%
  filter(pc_diff >= 1) %>%
  ggplot() +
  geom_histogram(bins = 40, aes(x=depth, fill=base)) +
  facet_grid(consensus_greater ~ region_name) +
  scale_fill_manual(values=palette_5) +
  labs(x = "Read Depth",
       y = "Count",
       title = "Read depth of frequency differences >= 1%") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))
# most differences at base depth < 1200


## Per sample distribution by nucleotide position
consensus_diff %>%
  filter(consensus_greater != "quasibam_sample") %>%
  mutate(greater_1_pc = if_else(pc_diff >= 1, 1, 0)) %>% 
  filter(quality_value > 1) %>%
  filter(source == "consensus") %>%
  group_by(Pos, region_name) %>%
  mutate(gt_1_pc = sum(100 * greater_1_pc) / n()) %>%
  ungroup() %>%
  group_by(Pos, region_name) %>%
  filter(pc_diff >= 1) %>%
  filter(pc_diff == max(pc_diff)) %>%  # Keep maximum diff per position, not per base & position
  ggplot() +
  geom_point(stat = "identity", aes(x = Pos, y = pc_diff, alpha = 0.2, 
                                    colour = gt_1_pc)) +
  scale_y_log10() +
  facet_wrap( ~ sample_name) + 
  scale_color_continuous(low = palette_5[1] , high = palette_5[5]) +
  guides(alpha = FALSE) +
  labs(x = "Nucleotide Position",
       y = "Maximum percent difference per base",
       colour = "Percent bases\nabove 1%") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))
# As before, sample 13 and 18 look particularily bad
# Also sample 3 has rare but very large differences in base frequency


```

## Number of positions comparison

```{r filtered_comparison, echo=FALSE}


# total number of bases
total_bases <- consensus_qb %>%
  group_by(region_name, sample_name) %>%
  summarise(max_length = max(Pos) * 4)  # as 4 bases

# Positions have at least less than 1% difference in base frequency
print("Positions with less than 1% difference")
consensus_diff %>%
  filter(quality_value > 1) %>%
  filter(source == "consensus") %>%
  group_by(Pos, region_name) %>%
  summarise(max_diff = max(pc_diff)) %>%
  filter(max_diff < 1) %>%
  group_by(region_name)  %>%
  summarise(number_of_positions = n())

# Positions have at least one base with > 1% difference
pass_filter <- consensus_diff %>%
  filter(pc_diff >= 1) %>%
  filter(quality_value > 1) %>%
  filter(source == "consensus") 

print("Positions with more than 1% difference")
pass_filter %>%
  group_by(Pos, region_name) %>%
  filter(Pos == max(Pos)) %>%
  group_by(region_name) %>%
  summarise(number_of_positions = n())

bases_pass <- pass_filter %>%
  filter(base != "Gap") %>%
  group_by(region_name) %>%
  summarise(number_of_bases = n())

tidy(prop.test(sum(bases_pass$number_of_bases), 
               sum(total_bases$max_length))) %>%
    select(conf.low:conf.high) 

# Positions have at least one base with > 5% difference
pass_filter <- consensus_diff %>%
  filter(pc_diff >= 5) %>%
  filter(quality_value > 1) %>%
  filter(source == "consensus") 

print("Positions with more than 5% difference")
pass_filter %>%
  group_by(Pos, region_name) %>%
  filter(Pos == max(Pos)) %>%
  group_by(region_name) %>%
  summarise(number_of_positions = n())

bases_pass <- pass_filter %>%
  filter(base != "Gap") %>%
  group_by(region_name) %>%
  summarise(number_of_bases = n())

  tidy(prop.test(sum(bases_pass$number_of_bases), 
                 sum(total_bases$max_length))) %>%
    select(conf.low:conf.high) 

# Positions have at least one base with > 10% difference
pass_filter <- consensus_diff %>%
  filter(pc_diff >= 10) %>%
  filter(quality_value > 1) %>%
  filter(source == "consensus") 

print("Positions with more than 10% difference")
pass_filter %>%
  group_by(Pos, region_name) %>%
  filter(Pos == max(Pos)) %>%
  group_by(region_name) %>%
  summarise(number_of_positions = n())

bases_pass <- pass_filter %>%
  filter(base != "Gap") %>%
  group_by(region_name) %>%
  summarise(number_of_bases = n())

tidy(prop.test(sum(bases_pass$number_of_bases), 
               sum(total_bases$max_length))) %>%
  select(conf.low:conf.high) 

pass_filter %>%
  group_by(Pos, region_name) %>%
  filter(pc_diff == max(pc_diff)) %>%
  select(Pos:region_name, Cons, depth:pc_diff, -source, base) %>%
  ungroup() %>%
  arrange(Pos, sample_name, region_name) %>%
  knitr::kable()

```

