---
title: "Pipeline comparison to consensus"
author: "Stef Piatek"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  html_document:
    keep_md: TRUE
params:
  pipeline: "vicuna_bwa"
  date_prefix: "171009"
---
Script to compare de novo pipeline quasibam to the frequency matrix from the FASTA pileup

Example usage from command line:

    Rscript -e "rmarkdown::render('scripts/quasibam-pipeline_comparison.Rmd', params = list(pipeline = 'vicuna_bwa', date_prefix = '170908'), 'html_document', '../reports/170908_vicuna-bwa_report.html')"

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(readr)
library(here)
library(glue)
library(dplyr)
library(tidyr)
library(stringr)
library(ggplot2)
library(broom)

# set up ggplot theme and palettes 
theme_custom <- function (base_size = 12, base_family = "") {
  theme_bw(base_size = base_size, base_family = base_family) %+replace%
    theme(
      axis.line.x = element_line(color = "black"),
      axis.line.y = element_line(color = "black"),
      panel.border = element_blank(),
      legend.key = element_rect(fill = NA, colour = NA),
      panel.grid.minor.y = element_line(colour="white"),
      panel.grid.major = element_line(colour = "white"),
      axis.title.x=element_text(vjust=-1.5),
      plot.margin= unit(c(0.5, 0, 0.5, 0), "cm")# top right bottom left
    )
}
theme_set(theme_custom())

palette_5 <- c("#FFA347","#142952", "#B80000","#97F297", "#000000")





```

## Comparison of quasibam to pipeline

- The consensus quasibam had been made from a set of FASTAs

    - FASTQs have been generated from these FASTAs
    - Mapped back to the consensus FASTA sequence 
    - quasi_bam ran on the resulting bam file


This document compares the consensus quasibam output to the pipeline.

- The NS3, NS5A and NS5B regions have been extracted from the files, discarding everything else


```{r load_data, warning=FALSE, message=FALSE}
# Load helper scripts and hcv reference information
source(here("scripts", "helper_functions.R"))
pipeline <- params$pipeline
print(glue("Pipeline analysed: {pipeline}"))
date_prefix <- params$date_prefix
print(glue("Analysing data from {date_prefix}"))
### Load data ----

samples <- list.files(here("data")) %>%
  str_extract(glue("({date_prefix}_[0-9]+)")) %>%
  unique() %>% 
  na.omit()

if(startsWith(pipeline, "iva")){
  samples <- samples[samples != "171009_6"]
}

# set up initial dataframe 
consensus_qb <- read_delim(file = here("data", glue("{samples[1]}_quasi_frequency_matrix.txt")),
                           delim = "\t") %>%
  mutate(sample_name = samples[1])%>%
  filter_to_roi()

pipeline_qb <- read_delim(file = here("data", glue("{samples[1]}_{pipeline}_quasibam.txt")),
                           delim = "\t") %>%
  select(-X22) %>%  # extra column added in import
  mutate(RefN = str_to_upper(RefN)) %>%
  mutate(sample_name = samples[1])%>%
  # align to regions of interest
  filter_to_roi() %>%
  # indel correction
  compensate_for_indels(sample = samples[1], ref_table = consensus_qb)

# add all other sets to the dataframe
for(item in 2:length(samples)){
  sample <- samples[item]
  consensus_qb <- read_delim(file = here("data", glue("{sample}_quasi_frequency_matrix.txt")),
                           delim = "\t") %>%
    mutate(sample_name = sample) %>%
    filter_to_roi() %>%
    bind_rows(consensus_qb)
  
  pipeline_qb <- read_delim(file = here("data", glue("{sample}_{pipeline}_quasibam.txt")),
                           delim = "\t") %>%
    select(-X22) %>%
    mutate(sample_name = sample) %>%
    mutate(RefN = str_to_upper(RefN)) %>%
    filter_to_roi() %>%
    compensate_for_indels(sample = sample, ref_table = consensus_qb) %>% 
    # 170908 where quasibam Cons is out of sync with RefN
    mutate(Pos = if_else(sample_name == "170908_18" & startsWith(pipeline, "vicuna"),
                         Pos -1,
                         as.double(Pos))) %>%
    mutate(RefN = if_else(sample_name == "170908_18" & startsWith(pipeline, "vicuna"),
                          lag(RefN, n = 1),
                          RefN)) %>%
    mutate(Pos = if_else(sample_name == "170908_3" & pipeline == "vicuna_smalt",
                         Pos -1,
                         as.double(Pos))) %>%
    mutate(RefN = if_else(sample_name == "170908_3" & pipeline == "vicuna_smalt",
                          lag(RefN, n = 1),
                          RefN)) %>%    
    mutate(Pos = if_else(sample_name == "170908_15" & pipeline == "iva_smalt",
                         Pos -1,
                         as.double(Pos))) %>%
    mutate(RefN = if_else(sample_name == "170908_15" & pipeline == "iva_smalt",
                          lag(RefN, n = 1),
                          RefN)) %>%   
    # 171009: 
    mutate(Pos = if_else(sample_name == "171009_11",
                         Pos -3,
                         as.double(Pos))) %>%
    mutate(RefN = if_else(sample_name == "171009_11",
                          lag(RefN, n = 3),
                          RefN)) %>%   
    filter(Pos >= 1) %>%
    bind_rows(pipeline_qb)
}

# change Positions to fit indel correction
pipeline_qb <- pipeline_qb %>%
  group_by(sample_name, region_name) %>%
  mutate(Pos = 1:n()) %>%
  ungroup()

```

## Size of sequence & Discordant basecalls



```{r discordant_basecalls}
### Bases removed from pipeline sequences 

if(nrow(indel_tbl) != 0){
  indel_tbl %>%
    group_by(sample_name, region_name, indel_type, Cons) %>%
    summarise(total_bases = n(),
              positions = glue("{min(Pos)} - {max(Pos)}"))
}



### Compare size of reference genome between consensus and pipeline
ref_length_consensus <- consensus_qb %>%
  group_by(sample_name, region_name) %>%
  summarise(ref_length = max(Pos))

pipeline_qb %>%
  group_by(sample_name, region_name) %>%
  summarise(ref_length = max(Pos)) %>%
  full_join(ref_length_consensus, by = c("sample_name", "region_name")
            , suffix = c("_pipeline", "_consensus")) %>%
  mutate(length_difference = ref_length_pipeline - ref_length_consensus) %>%
  filter(length_difference != 0) %>%
  arrange(sample_name)

### Compare data where basecalls are discordant within the sample and ref ----

discordant_basecalls <- pipeline_qb %>%
  filter(RefN != Cons) %>%
  filter(!(Cons == "R" & str_detect(RefN, "[AG]"))) %>%
  filter(!(Cons == "Y" & str_detect(RefN, "[CT]"))) %>%
  filter(!(Cons == "S" & str_detect(RefN, "[GC]"))) %>%
  filter(!(Cons == "W" & str_detect(RefN, "[AT]"))) %>%
  filter(!(Cons == "K" & str_detect(RefN, "[GT]"))) %>%
  filter(!(Cons == "M" & str_detect(RefN, "[AC]"))) %>%  
  select(Pos:Gap, Cons, sample_name, region_name) %>%
  filter(Depth != 0) %>%
  arrange(sample_name, region_name, Pos) 

discordant_basecalls %>%
  group_by(region_name, sample_name, Cons) %>%
  summarise(total = n(),
            positions = glue("{min(Pos)} - {max(Pos)}"))

```

## Basecall frequency differences

### Parse data for plotting


```{r merge_quasibams}


tmp_consensus_diff <- consensus_qb %>%
  full_join(pipeline_qb, by=c("Pos", "sample_name", "region_name"), 
            suffix=c("_consensus", "_pipeline")) %>%
  # Keep only pipeline depth as depth
  mutate(depth = Depth_pipeline) %>%
  select(- starts_with("Depth_")) %>%
  # tidy data to have base, sample and value columns
  gather(key = base_source, value = value, 
         starts_with("A_"), starts_with("C_"), starts_with("T_"), starts_with("G_"), 
         starts_with("Gap_")) %>%
  separate(col = base_source, into = c("base", "source"), sep = "_" ) %>%
  # get difference in percent
  group_by(Pos, base, sample_name, region_name) %>%
  arrange(source) %>%
  mutate(pc_diff = max(value) - min(value)) %>%
  # plot using log, so no negative. use factor to determine which value is greater
  mutate(consensus_greater = if_else(value > lead(value),
                                     "consensus value greater",
                                     "consensus value smaller",
                                     missing="pipeline_sample")) %>%
  mutate(consensus_greater = parse_factor(consensus_greater, 
                                          levels = c("consensus value greater",
                                                     "consensus value smaller",
                                                     "pipeline_sample"))) %>%
  ungroup() %>%
  mutate(sample_name = if_else(str_detect(sample_name, "([0-9]{6}_[0-9]{1}$)"),
                               str_replace(sample_name, "_", "_0"),
                               sample_name)) 

# reshape quality data into long format and merge back into consensus_diff
quality_values <- tmp_consensus_diff %>%
  select(Pos, starts_with("q"), sample_name, region_name) %>%
  gather(key = base, value = quality_value, 
         starts_with("q")) %>%
  mutate(base = str_replace(base, "q", "")) %>%
  group_by(Pos, base, sample_name, region_name) %>%
  dplyr::slice(1) %>%
  ungroup()


consensus_diff <- tmp_consensus_diff %>%
  select(-starts_with("q")) %>%
  full_join(quality_values, by=c("Pos", "base", "sample_name", "region_name")) %>%
  mutate(base = parse_factor(base, levels=c("A", "C", "G", "T", "Gap"))) %>%
  # for non-called bases, change base quality to NA
  mutate(quality_value = replace(quality_value,
                                 quality_value == 0 & value == 0,
                                 NA))
# remove temporary data
rm("quality_values")
rm("tmp_consensus_diff")

```

### Plotting base frequency differences

- Maximum base frequency difference is 4%, the marjority is under 1%.
- largest peak in differences in base frequency is at ~ 1,500 nt, where a gap in the reference is found

    - Some of the base frequency differences at this position has a depth of 0 in the pipeline sample

- Mapping quality of the base is high and does not change with the frequency difference

```{r plots}
### plot per base differences ----
pipeline_qb %>%
  group_by(Pos, region_name) %>%
  filter(Depth > 100) %>%
  summarise(percent_covered = 100 * n() / length(samples)) %>%
  arrange(region_name) %>%
  ggplot() +
  geom_point(aes(x = Pos, y = percent_covered, colour = '>= 100')) + 
  scale_x_continuous(limits=c(1, max(consensus_qb$Pos))) +
  scale_colour_manual(values=palette_5) +
  facet_wrap(~region_name) + 
  labs(x = "Nucleotide position",
       y = "Percent of samples",
       title = "Percent of samples with depth above 100",
       colour = "Depth")

# double check no missed rows
pipeline_qb %>%
  group_by(Pos, region_name) %>%
  summarise(percent_covered = 100 * n() / length(samples)) %>%
  group_by(region_name) %>%
  arrange(region_name) %>%
  filter(lag(Pos) != Pos - 1 | lead(Pos) != Pos + 1)

## Distribution of frequency differences per base ----
consensus_diff %>%
  filter(consensus_greater != "pipeline_sample") %>%
  filter(pc_diff > 0.1) %>%
  ggplot(aes(x=pc_diff, fill=base)) +
  geom_histogram(bins=40) +
  scale_x_log10() + 
  scale_y_log10() + 
  facet_grid(consensus_greater ~ region_name) + 
  scale_fill_manual(values=palette_5) +
  labs(x = "Frequency difference",
       y = "Count",
       title = "Differences compared to consensus")
# 


## Distribution of frequency differences per sample ----
consensus_diff %>%
  filter(consensus_greater != "pipeline_sample") %>%
  filter(pc_diff > 0.1) %>%
  ggplot(aes(x = pc_diff, fill = region_name)) +
  geom_histogram(bins=40) +
  scale_x_log10() + 
  scale_y_log10() + 
  scale_fill_manual(values = palette_5) +
  facet_wrap(~ sample_name) + 
  labs(x = "Frequency difference",
       y = "Count",
       title = "Differences compared to consensus")

## Per position frequency differences by sample ----
consensus_diff %>%
  filter(consensus_greater != "pipeline_sample") %>%
  filter(pc_diff >= 1) %>%
  ggplot(aes(x = Pos, fill = region_name)) +
  geom_histogram(bins=40) +
  scale_fill_manual(values=palette_5) +
  facet_wrap(~ sample_name, scales = "free_y") + 
  labs(x = "Nucleotide Position",
       y = "Count",
       fill = "Region name",
       title = "Number of differences >1% by position") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))


## proportion of samples with base difference at position ----
consensus_diff %>%
  filter(consensus_greater != "pipeline_sample") %>%
  mutate(greater_1_pc = if_else(pc_diff >= 1, 1, 0)) %>% 
  group_by(Pos, base, source, region_name) %>%
  summarise(gt_1_pc = sum(100 * greater_1_pc) / n(),
            max_pc_diff = max(pc_diff)) %>%
  filter(max_pc_diff >= 1) %>%
ggplot(aes(x = max_pc_diff, y = gt_1_pc, colour = base, alpha = 0.2)) +
  geom_jitter() +
  scale_x_log10() + 
  facet_grid(region_name ~ base) + 
  scale_colour_manual(values=palette_5) +
  guides(alpha=FALSE) + 
  labs(x = "Maximum difference in base frequency",
       y = "Percent of samples",
       title = "Percent of samples with >1% difference in base frequency")


## Distribution of differences greater than 1% by reference position ----
consensus_diff %>%
  # remove samples with frequencies within 1% of eachother 
  filter(pc_diff >= 1) %>%
  filter(consensus_greater != "pipeline_sample") %>%
  ggplot() +
  geom_histogram(bins=40, aes(x=Pos, fill=base)) +
  facet_grid(consensus_greater ~ region_name) + 
  scale_fill_manual(values=palette_5) +
  labs(x = "Nucleotide position",
       y = "Count",
       title = "Differences by position, over 1%") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))


## Read depth of reads with frequency differences greater than 1%
consensus_diff %>%
  filter(consensus_greater != "pipeline_sample") %>%
  filter(pc_diff >= 1) %>%
  ggplot() +
  geom_histogram(bins = 40, aes(x=depth, fill=base)) +
  facet_grid(consensus_greater ~ region_name) +
  scale_fill_manual(values=palette_5) +
  labs(x = "Read Depth",
       y = "Count",
       title = "Read depth of frequency differences >= 1%") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))

## Per sample distribution by nucleotide position
consensus_diff %>%
  filter(source == "pipeline") %>% 
  mutate(greater_1_pc = if_else(pc_diff >= 1, 1, 0)) %>% 
  filter(quality_value > 1) %>%
  group_by(Pos, region_name) %>%
  mutate(gt_1_pc = sum(100 * greater_1_pc) / n()) %>%
  filter(pc_diff >= 1) %>%
  filter(pc_diff == max(pc_diff)) %>%  # Keep maximum diff per position, not per base & position
  ggplot() +
  geom_point(stat = "identity", aes(x = Pos, y = pc_diff, alpha = 0.2, 
                                    colour = gt_1_pc)) +
  scale_y_log10() +
  facet_wrap( ~ sample_name) + 
  scale_color_continuous(low = palette_5[1] , high = palette_5[5]) +
  guides(alpha = FALSE) +
  labs(x = "Nucleotide Position",
       y = "Maximum percent difference per base",
       colour = "Percent bases\nabove 1%") +
  theme(axis.text.x  = element_text(angle=45, vjust=0.5))



```

## Number of positions comparison

- for 171009, samples 10, 11 and 12 have been filtered out

```{r filtered_comparison, echo=FALSE}

# total number of bases
total_bases <- consensus_qb %>%
#  filter(!sample_name %in% c('171009_10', '171009_11', '171009_12')) %>%
  group_by(region_name, sample_name) %>%
  summarise(max_length = max(Pos) * 4)  # as 4 bases

total_bases_filtered <- consensus_qb %>%
  filter(!sample_name %in% c('171009_10', '171009_11', '171009_12')) %>%
  group_by(region_name, sample_name) %>%
  summarise(max_length = max(Pos) * 4)  # as 4 bases

analyse_threshold(1, "pipeline")
position_diff_export %>% knitr::kable()
conf_int_export %>% knitr::kable()

analyse_threshold(5, "pipeline")
position_diff_export %>% knitr::kable()
conf_int_export %>% knitr::kable()

analyse_threshold(10, "pipeline")
position_diff_export %>% knitr::kable()
conf_int_export %>% knitr::kable()

pass_filter %>%
  filter(!sample_name %in% c('171009_10', '171009_11', '171009_12')) %>%
  group_by(Pos, region_name) %>%
  filter(pc_diff == max(pc_diff)) %>%
  select(Pos:region_name, Cons, depth:pc_diff, -source, base) %>%
  ungroup() %>%
  arrange(Pos, sample_name, region_name) %>%
  knitr::kable()


```

